ext/phar/phar_object.c

```diff
@@ -2074,18 +2075,33 @@ static zend_object *phar_rename_archive(phar_archive_data *phar, char *ext, zend
 	}
 
 	oldpath = estrndup(phar->fname, phar->fname_len);
-	if ((oldname = zend_memrchr(phar->fname, '/', phar->fname_len))) {
-		++oldname;
-	} else {
-		oldname = phar->fname;
-	}
+	oldname = zend_memrchr(phar->fname, '/', phar->fname_len);
+	++oldname;
 	oldname_len = strlen(oldname);
 
 	basename = estrndup(oldname, oldname_len);
-	spprintf(&newname, 0, "%s.%s", strtok(basename, "."), ext);
-	efree(basename);
+	
+	/* find extension name: .phar suffixed with anything and .zip* and .tar* without .phar for data archives */
+	oldname_ext = strstr(basename, ".phar");
+	if (oldname_ext == NULL) {
+		oldname_ext = strstr(basename, ".tar");
+		if (oldname_ext == NULL) {
+			oldname_ext = strstr(basename, ".tgz");
+			if (oldname_ext == NULL) {
+				oldname_ext = strstr(basename, ".zip");
+			}
+		}
+	}
+
```

Comment:
You could make the code be far less indented and more readable and more maintainable by using something like:
```
    int i;
    char *extensions[] = {
        ".phar",
        ".tar",
        ".tgz",
        ".zip"
    };
    
    int extension_count = sizeof(extensions)/sizeof(extensions[0]);

    for (i=0 ; oldname_ext == NULL && i < extension_count ; i++){
        oldname_ext = strstr(basename, extensions[i]);
    }

```